### PR TITLE
Fix recovery in datasets

### PR DESCRIPTION
**Pull request by MaxiBoether**
Before, we had recovery logic based on reply indices. However, while working on storage, I realized those responses come non deterministically from multiple threads. Hence, we cannot rely on the ordering. We need to keep track of the sample IDs we already yielded. I changed the logic to just keep a list which is cheap to append to, and only convert to a set / hash table as soon as we failed once and we actually need to do many `in` checks.

### CODE REVIEW

File: modyn/evaluator/internal/dataset/evaluation_dataset.py [hunk 1]
```diff
@@ -125,15 +126,24 @@ def _get_keys_from_storage(self, worker_id: int, total_workers: int) -> Iterable
...
+                            processed_keys.update(resp_keys.keys)
+                            yield new_keys
 
                 except grpc.RpcError as e:
+                    has_failed = True
+                    # Convert processed keys to set on first failure
+                    processed_keys = set(processed_keys) if isinstance(processed_keys, list) else processed_keys
+
                     self._info(
-                        "gRPC error occurred, last index = " + f"{last_processed_index}: {e.code()} - {e.details()}",
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/evaluator/internal/dataset/evaluation_dataset.py:136]: This logic was here before the PR, anyways, the reason we reraise the error is tenacity. We are using tenacity retrying logic here. We just have a try-catch to intercept the exception and log some information, then we raise the error again and tenacity takes care of that. In the PR that introduced this were unit tests, at least for the trainer server dataset

[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/evaluator/internal/dataset/evaluation_dataset.py:136]: This PR fixes the logic on what happens on an exception, basically

[the remark this answers is not in the document]
Comment (XianzheMa) [reply modyn/evaluator/internal/dataset/evaluation_dataset.py:136]: Ah okay

**Comment by XianzheMa** [comment]
Can you piggyback deleting this log https://github.com/eth-easl/modyn/blob/55fef7c88b66a61d375d1555db46682157f3bbf5/modyn/supervisor/internal/pipeline_executor/pipeline_executor.py#L480 in this PR? This log is basically wrong because `len(s.triggers)` is not the number of triggers we this time are processing.

**Review by XianzheMa — APPROVED** [review]

**Pull request merged**